src/all/exhentai/src/eu/kanade/tachiyomi/extension/all/exhentai/EXHentai.kt

```diff
@@ -0,0 +1,654 @@
+import okhttp3.CacheControl
+import okhttp3.CookieJar
+import okhttp3.Headers
+import okhttp3.Request
+import okhttp3.Response
+import org.jsoup.nodes.Element
+import rx.Observable
+import uy.kohesive.injekt.Injekt
+import uy.kohesive.injekt.api.get
+import java.net.URLEncoder
+
+abstract class EXHentai(
+    override val lang: String,
+    private val ehLang: String,
+) : ConfigurableSource, HttpSource() {
+
+    private val preferences: SharedPreferences by lazy {
+        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
+    }
+    private val webViewCookieManager: CookieManager by lazy { CookieManager.getInstance() }
+    override val name = "EXHentai"
+
+    private val memberId: String = getMemberIdPref()
+    private val passHash: String = getPassHashPref()
+
+    override val baseUrl: String
+        get() = if (memberId.isEmpty() || passHash.isEmpty()) {
+            "https://forums.e-hentai.org/index.php?act=Login"
+        } else {
+            "https://exhentai.org"
+        }
+
+    override val supportsLatest = true
+
+    private var lastMangaId = ""
+
+    // true if lang is a "natural human language"
+    private fun isLangNatural(): Boolean = lang !in listOf("none", "other")
+
+    private fun genericMangaParse(response: Response): MangasPage {
+        val doc = response.asJsoup()
+        val mangaElements = doc.select("table.itg td.glname")
+            .let { elements ->
+                if (isLangNatural() && getEnforceLanguagePref()) {
+                    elements.filter { element ->
+                        // only accept elements with a language tag matching ehLang or without a language tag
+                        // could make this stricter and not accept elements without a language tag, possibly add a sharedpreference for it
+                        element.select("div[title^=language]").firstOrNull()?.let { it.text() == ehLang } ?: true
+                    }
+                } else {
+                    elements
+                }
+            }
+        val parsedMangas: MutableList<SManga> = mutableListOf()
+        for (i in mangaElements.indices) {
+            val manga = mangaElements[i].let {
+                SManga.create().apply {
+                    // Get title
+                    it.select("a")?.first()?.apply {
+                        title = this.select(".glink").text()
+                        url = ExGalleryMetadata.normalizeUrl(attr("href"))
+                        if (i == mangaElements.lastIndex) {
+                            lastMangaId = ExGalleryMetadata.galleryId(attr("href"))
+                        }
+                    }
+                    // Get image
+                    it.parent()?.select(".glthumb img")?.first().apply {
+                        thumbnail_url = this?.attr("data-src")?.takeIf { it.isNotBlank() } ?: this?.attr("src")
+                    }
+                }
+            }
+            parsedMangas.add(manga)
+        }
```

Comment:
`mangaElements.map { ...`

Reply:
Might need `.mapIndexed`, since there's a condition checking if it's the last element.